src/AspNet.Security.OAuth.Untappd/UntappdAuthenticationHandler.cs

```diff
@@ -29,6 +33,63 @@ public class UntappdAuthenticationHandler : OAuthHandler<UntappdAuthenticationOp
         {
         }
 
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
+        {
+            var tokenRequestParameters = new Dictionary<string, string>()
+            {
+                { "client_id", Options.ClientId },
+                { "redirect_uri", context.RedirectUri },
+                { "client_secret", Options.ClientSecret },
+                { "code", context.Code },
+                { "grant_type", "authorization_code" },
+            };
+
+            // PKCE https://tools.ietf.org/html/rfc7636#section-4.5, see BuildChallengeUrl
+            if (context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier))
+            {
+                tokenRequestParameters.Add(OAuthConstants.CodeVerifierKey, codeVerifier!);
+                context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+            }
+
+            var requestContent = new FormUrlEncodedContent(tokenRequestParameters!);
+
+            var builder = new UriBuilder(Options.TokenEndpoint);
+            builder.Port = -1;
+            var query = HttpUtility.ParseQueryString(builder.Query);
+            query["client_id"] = Options.ClientId;
+            query["redirect_uri"] = context.RedirectUri;
+            query["client_secret"] = Options.ClientSecret;
+            query["code"] = context.Code;
+            builder.Query = query.ToString();
+            string url = builder.ToString();
+
+            using var requestMessage = new HttpRequestMessage(HttpMethod.Get, url);
+            requestMessage.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            requestMessage.Content = requestContent;
+            requestMessage.Version = Backchannel.DefaultRequestVersion;
+
+            var response = await Backchannel.SendAsync(requestMessage, Context.RequestAborted);
+            if (response.IsSuccessStatusCode)
+            {
+                var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted));
+                return OAuthTokenResponse.Success(payload);
+            }
+            else
+            {
+                var error = "OAuth token endpoint failure: " + await Display(response);
+                return OAuthTokenResponse.Failed(new Exception(error));
+            }
+        }
```

Comment:
Please use the same failure handling pattern as our other providers. Example:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/9e1fef49de1abd808d2d6cfde26a1c66ce51c6da/src/AspNet.Security.OAuth.GitHub/GitHubAuthenticationHandler.cs#L86-L95